Update Spark Operator getting started guide. (#2311)
NVIDIA/spark-rapids: docs/get-started/getting-started-kubernetes.md

sc.stop()
```

## Running Spark Applications using Spark Operator

Using Spark Operator is another way to submit Spark Applications into a Kubernetes Cluster.

> 1. Copy Spark Application jars/files into the docker image when preparing docker image.

Reviewer: so is there any difference between this and the regular instructions for k8s?  either way we make the jars available in the docker image and set the configs appropriately.  Is the docker image different or can you use the same one?

Author: Correct. as per the Spark Operator user guide : "A SparkApplication should set .spec.deployMode to cluster, as client is not currently implemented. The driver pod will then run spark-submit in client mode internally to run the driver program. "

Author: @tgravescs  i just added a new sentence
```
This is because currently only cluster deployment mode is supported by Spark Operator.
```

Reviewer: I'm not sure that addresses my concern, I was thinking more along the lines of the docker image. Above we state exactly how to create it and here we just say copy it into the docker image. If the docker image is created the exact same way, perhaps we just reference that section.

My other thought was actually do we even want this here at all?  It seems like a standard spark operator way of doing things so if there is nothing special do we want to maintain our own instructions?

Author: I see. Actually if we modify the initial dockerfile to put the test.py inside, then yes, it is the same docker image as previous examples, no difference. 
Since users may not need that test.py for their real production image, so we did not COPY it in the previous example.

IMO, the extra steps for "Spark Operator"'s most important goal is to provide an YAML file example. 
Other that, I did not put detailed steps on how to install "Spark Operator" here, and just point it to the external links.

Reviewer: ok so I think we should clarify the docker image part. The docker has preparing the docker image at the top so I take that applies to all sections, but here we say again copy the jars to it, but we have already done that. 

Perhaps we need to split this into separate sub-sections. Like submitting with spark cli, submitting with spark operator and those are sections after the docker image creation, which is prerequisite.

Author: I just added an line when preparing the docker image saying that 
```
You can modify the Dockerfile to copy your application into the docker image, i.e. test.py.
```

Then now I changed it to "locate" the application file/jar instead of COPY.
Here we assume the test.py is the user's application file which is already copied into the docker image.

For example, we copy above `test.py` in directory `/opt/sparkRapidsPlugin` inside the docker image.

2. Create spark operator using `helm`.

Follow [Spark Operator quick start guide](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/quick-start-guide.md)

3. Create a Spark Application YAML file

> For example, below file named `testpython-rapids.yaml` is created:

Reviewer 2: ```suggestion
For example, create a file named `testpython-rapids.yaml` with the following contents:
```

```
apiVersion: "sparkoperator.k8s.io/v1beta2"
kind: SparkApplication
metadata:
  name: testpython-rapids
  namespace: default
spec:
  sparkConf:
    "spark.ui.port": "4045"
    "spark.rapids.sql.concurrentGpuTasks": "1"
    "spark.executor.resource.gpu.amount": "1"

>     "spark.task.resource.gpu.amount": "1"

Reviewer 2: I realize this is done elsewhere in this file and not just here, but do we really want all these examples to setup an executor that can only run 1 task concurrently?

Author: I checked all previous examples and all are using 1 task/executor. Since this is for examples, I am fine. 
But if we have this concern, we can point to them to our tuning guide saying something like "For more parameter tuning related topic, refer to this tuning guide." How do you think?

Reviewer 2: It doesn't have to be addressed in this PR, I'm more wondering in general if it's a good idea for these examples to have such a suboptimal setup without calling it out as such.

Author: I recalled one reason why we put minimum resource here is sometimes people are testing a single node K8s cluster with 1 GPU inside. This can at least run through without much troubleshooting done to figure out GPU resource not enough issue.

Reviewer: 1 is the default so we could just leave it off.

Reviewer 2: but I think that's even worse, right?  We don't want only one task running per executor, we would normally want users to set this to a fraction of a GPU.  I'm fine with this going in as-is since it's consistent with the other examples, but I think there should be a followup to address the very suboptimal config especially if the query doesn't run fully on the GPU.

Reviewer: Why is that worse?  Our defaults should be something we think are a good starting point, otherwise why is it the default? If you are saying that we should put it in at some other value for most people to start with, we should just change the default and not specify the config explicitly.

Reviewer: Oh sorry, I think I was talking bout different configs, I you meant the spark.task.resource.gpu.amount, then agree, setting this lower in example is good.  I was referring to spark.rapids.sql.concurrentGpuTasks

Reviewer 2: I mean it's worse in the sense that it's hidden.  The default makes sense from a safety, it-will-work-if-anything-works perspective, but it's going to be very inefficient for many real-world queries.  My concern is users will set this up per the instructions, get something to run, see the performance is terrible, and stop there.

Like I said above, I'm totally cool with leaving it in (or removing it since it's the default), but we should mention in htis document that the setup is very much untuned/suboptimal and point them to the tuning guide.  The tuning guide needs to discuss setting this config, as it currently doesn't mention it at all.

Reviewer 2: > you meant the spark.task.resource.gpu.amount

Yes, that's what I'm worried about.  Running 1 task per executor and leaving a ton of CPU cores completely unused is not going to be stellar performance for legacy shuffle or other parts of the query that are not GPU-bound.

I'll file a followup issue for addressing this, as the tuning guide really needs to address this and that's outside the scope of this PR.

Reviewer 2: Filed #2328 to track updating the tuning guide and mentioned updating the examples in other docs, so we don't need to deal with that particular issue in this PR.

    "spark.executor.resource.gpu.discoveryScript": "/opt/sparkRapidsPlugin/getGpusResources.sh"
    "spark.executor.resource.gpu.vendor": "nvidia.com"
    "spark.executor.extraClassPath": "/opt/sparkRapidsPlugin/rapids-4-spark.jar:/opt/sparkRapidsPlugin/cudf.jar"
    "spark.driver.extraClassPath": "/opt/sparkRapidsPlugin/rapids-4-spark.jar:/opt/sparkRapidsPlugin/cudf.jar"
  type: Python
  pythonVersion: 3
  mode: cluster
  image: "<IMAGE_NAME>"
  imagePullPolicy: Always
  mainApplicationFile: "local:///opt/sparkRapidsPlugin/test.py"
  sparkVersion: "3.1.1"
  restartPolicy:
    type: Never
  volumes:
    - name: "test-volume"
      hostPath:
        path: "/tmp"
        type: Directory
  driver:
    cores: 1
    coreLimit: "1200m"
    memory: "1024m"
    labels:
      version: 3.1.1
    serviceAccount: spark
    volumeMounts:
      - name: "test-volume"
        mountPath: "/tmp"
  executor:
    cores: 1
    instances: 1
    memory: "5000m"
    gpu:
      name: "nvidia.com/gpu"
      quantity: 1
    labels:
      version: 3.1.1
    volumeMounts:
      - name: "test-volume"
        mountPath: "/tmp"
```

4. Submit the Spark Application

```
sparkctl create testpython-rapids.yaml
```

> Note: sparkctl can be built from the Spark Operator repo after [installing golang](https://golang.org/doc/install):

Reviewer 2: ```suggestion
Note: `sparkctl` can be built from the Spark Operator repo after [installing golang](https://golang.org/doc/install):
```

```
cd sparkctl
go build -o sparkctl
```

5. Check the driver log

```
sparkctl log testpython-rapids
```

6. Check the status of this Spark Application

```
sparkctl status testpython-rapids
```

7. Port forwarding when Spark driver is running

```
sparkctl forward testpython-rapids --local-port 1234 --remote-port 4045
```

Then open browser with `http://localhost:1234/` to check Spark UI.

8. Delete the Spark Application

```
sparkctl delete testpython-rapids
```

> 
> 

Reviewer 2: ```suggestion
```
